lib/linkerd2-trace-context/src/layer.rs

```diff
@@ -0,0 +1,184 @@
+{
+    type Response = Service<Make::Response, Sink>;
+    type Error = Make::Error;
+    type Future = MakeFuture<Make::Future, Sink>;
+
+    fn poll_ready(&mut self) -> Poll<(), Make::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, target: Target) -> Self::Future {
+        let inner = self.inner.call(target);
+
+        MakeFuture {
+            inner,
+            sink: Some(self.sink.clone()),
+        }
+    }
+}
+
+// === impl MakeFuture ===
+
+impl<Fut: Future, Sink> Future for MakeFuture<Fut, Sink> {
+    type Item = Service<Fut::Item, Sink>;
+    type Error = Fut::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = try_ready!(self.inner.poll());
+        let sink = self.sink.take().expect("poll called after ready");
+        Ok(Async::Ready(Service { inner, sink }))
+    }
+}
+
+// === impl Service ===
+
+impl<Svc, Body, Sink> tower::Service<http::Request<Body>> for Service<Svc, Sink>
+where
+    Svc: tower::Service<http::Request<Body>>,
+    Sink: SpanSink + Clone,
+{
+    type Response = Svc::Response;
+    type Error = Svc::Error;
+    type Future = futures::future::Either<Svc::Future, SpanFuture<Svc::Future, Sink>>;
+
+    fn poll_ready(&mut self) -> Poll<(), Svc::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut request: http::Request<Body>) -> Self::Future {
+        let mut trace_context = propagation::unpack_trace_context(&request);
+        let mut path = None;
+
+        if let Some(ref mut context) = trace_context {
+            trace!(message = "got trace context", ?context);
+            propagation::increment_span_id(&mut request, context);
+            // If we plan to sample this span, we need to copy the path from
+            // the request before dispatching it to inner.
+            if context.is_sampled() {
+                path = Some(request.uri().path().to_string());
+            }
+        }
+
+        let f = self.inner.call(request);
+
+        if let Some(propagation::TraceContext {
+            propagation: _,
+            version: _,
+            trace_id,
+            parent_id,
+            flags,
+            span_id: Some(span_id),
+        }) = trace_context
+        {
+            if flags.is_sampled() {
+                trace!(message = "span will be sampled", ?span_id);
+                let span = Span {
+                    trace_id,
+                    span_id,
+                    parent_id,
+                    span_name: path.unwrap_or_else(String::new),
+                    start: SystemTime::now(),
+                    // End time will be updated when the span completes.
+                    end: SystemTime::now(),
+                };
+                let span_fut = SpanFuture {
+                    span: Some(span),
+                    inner: f,
+                    sink: self.sink.clone(),
+                };
+                return futures::future::Either::B(span_fut);
+            }
```

Comment:
> This all seems a bit tortured to me... 
This is how I always feel

I'm not sure I understand your suggestion, though.  We can't send the span in `call` because we don't know the span end time.  In a subsequent PR I'll be adding some metadata from the response to the span as well, which is not available until the future completes.

Reply:
Ah, OK. Makes sense

I would probably personally opt to have a ResponseFuture type that hides the messy innards (basically, holding an `Option<SendSpan>`, but this is okay too.

But, please for the love of all things good import `Either` so we don't have to look at the `futures::future::` every time